sprint 7

PrettyDG opened the pull request:
Добавлен новый класс 'FileBackedTaskManager'. Он повторяет методы 'InMemoryTaskManager', но сохраняет результат в файл. Также реализована возможность добавления задач через файл

PrettyDG committed:
Добавлен новый класс 'FileBackedTaskManager'. Он повторяет методы 'InMemoryTaskManager', но сохраняет результат в файл. Также реализована возможность добавления задач через файл

PrettyDG committed:
Улучшение для прохождение тестов GitHub #1

PrettyDG commented:
Вновь не проходят тесты из-за того, что нет папки "test". Но она присутствует

avfyodorov commented on src/controllers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,132 @@
+
+import java.io.*;
+
+public class FileBackedTaskManager extends InMemoryTaskManager implements TaskManager {
```
implements TaskManager {

Повторно реализовывать интерфейс не нужно, потому что он уже реализован в классе предке.

avfyodorov commented on src/controllers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,132 @@
+
+public class FileBackedTaskManager extends InMemoryTaskManager implements TaskManager {
+
+    File file;
```
File file;

Для соблюдения инкапсуляции переменную лучше сделать приватной

avfyodorov commented on src/controllers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,132 @@
+        this.file = file;
+    }
+
+    public FileBackedTaskManager loadFromFile(File file) {
```
не хватает static

Нужно изменить спецификацию метода, чтобы было соответствие с ТЗ.
 создайте статический метод static FileBackedTasksManager loadFromFile(File file), который будет восстанавливать данные менеджера из файла при запуске программы

avfyodorov commented on src/controllers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    public FileBackedTaskManager loadFromFile(File file) {
+        FileBackedTaskManager manager = new FileBackedTaskManager(new File("file.txt"));
```
File("file.txt"

Лучше было бы перенести этот файл из корня проекта в отдельный подкаталог с ресурсами.

avfyodorov commented on src/controllers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,132 @@
+            }
+        } catch (FileNotFoundException exception) {
+            System.out.println("Файл не найден");
+        } catch (IOException exception) {
```
Тут немного   не так. Перехватываете IOException и сами пробрасываете
 throw new ManagerSaveException(«Ошибка и т.д.."); 
Т.е. этот (первый) catch не нужен.

avfyodorov commented on src/controllers/ManagerSaveException.java:
```diff
@@ -0,0 +1,9 @@
+
+import java.io.IOException;
+
+public class ManagerSaveException extends IOException {
```
Лучше наследовать от RuntimeException
Из ТЗ:
кидать собственное непроверяемое исключение ManagerSaveException. Благодаря этому можно не менять сигнатуру методов интерфейса менеджера.

public class RuntimeException extends Exception — базовый класс для ошибок во время выполнения. Относится к необрабатываемым исключениям (uncatched\unchecked). Как сказано в описании класса — это суперкласс, исключения которого могут быть выброшены во время нормальной работы JVM.

avfyodorov commented on src/controllers/ManagerSaveException.java:
```diff
@@ -0,0 +1,9 @@
+package controllers;
```
Собственные исключения принято хранить в отдельном пакете, например exceptions

avfyodorov commented on src/controllers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    @Override
+    public DefaultTask createDefaultTask(DefaultTask currentTask) throws ManagerSaveException {
```
throws ManagerSaveException

Здесь и дальше. 
Если изменить наследование собственного исключения, то данное ключевое слово будет лишним.

avfyodorov commented on src/controllers/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,132 @@
+                    throw new ManagerSaveException("Произошла ошибка сохранения файла!");
+                }
+            }
+        } catch (IOException exception) {
```
Вот здесь Совершенно верно сделано, точно так же нужно организовать и при загрузке данных.

avfyodorov commented:
> Вновь не проходят тесты из-за того, что нет папки "test". Но она присутствует


Присутствует-то она, конечно, присутствует... но только не на месте и не помечена как тестовый пакет.  
Надо будет реструктуризировать проект таким образом, чтобы дерево пакетов выглядело приблизительно так, как показано на рисунке. Делается это несложно, через рефакторинг F6. Недостающие пакеты создаются вручную, ещё нужно не забыть определить пакет test как корень тестовых пакетов. 
Я думаю, это должно помочь. 
И ещё один момент-нужно добавить тестов для проверки загрузки и сохранения данных в файл. 

![2024-09-28_14-47-53](https://github.com/user-attachments/assets/8310f7ff-1198-439f-ad1b-9765ea3c2e88)

avfyodorov reviewed (commented):
Добрый день, Данила! 

Есть несколько уточнений По проекту.

На Ваш вопрос ответил и гитхабе и в пачке, Посмотрите, пожалуйста. 

Ещё один момент-в структуре проекта множество ссылок на библиотеку JUnit, нужно почистить, оставить только одну, 5-ую.

PrettyDG committed:
Создал отдельные подкаталоги для ресурсов и собственных исключений.
Добавил дополнительные тесты для 'FileBackedTaskManager'.
Немного переработал метод 'fromStringToTask'

PrettyDG commented:
Добрый день! Благодарю за подсказку о пометке тестовых пакетов. Тесты GitHub наконец полностью засчитаны. 

В проект добавил тесты для нового класса 'FileBackedTaskManagerTest'. Все тесты проходят верно если их запускать вместе с предыдущим 'InMemoryTaskManagerTest', однако по-отдельности перестают работать. Нужно ли как-то это исправлять, чтобы они и отдельно друг от друга проходили?

PrettyDG commented:
Насчёт ссылок на библиотеку JUnit не особо понял, где именно их необходимо почистить

avfyodorov commented:
> Добрый день! Благодарю за подсказку о пометке тестовых пакетов. Тесты GitHub наконец полностью засчитаны.
> 
> В проект добавил тесты для нового класса 'FileBackedTaskManagerTest'. Все тесты проходят верно если их запускать вместе с предыдущим 'InMemoryTaskManagerTest', однако по-отдельности перестают работать. Нужно ли как-то это исправлять, чтобы они и отдельно друг от друга проходили?

Вообще говоря, Желательно, чтобы тест файлового менеджера проходил сам по себе, но сейчас задерживаться на этом не будем. 
Будет время, посмотрите, пожалуйста.

avfyodorov commented:
> Насчёт ссылок на библиотеку JUnit не особо понял, где именно их необходимо почистить

Вот здесь в прошлый раз было семь-восемь разных библиотек. Сейчас всё в порядке, они убежали. 
![2024-09-30_12-11-59](https://github.com/user-attachments/assets/13a24553-059a-49e3-a94c-ff9e642da6ff)

avfyodorov reviewed (commented):
Добрый день, Данила!

Замечаний нет.
Работа принята.
